Refactor JS Function add_tag() and create add_comment() for Javascript API

nstjean opened the pull request:
Fixes #6411  (<=== Add issue number here)

Creating a Pull Request early to get feedback! I'm not sure if I'm doing this right.



Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [ ] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

nstjean commented:
@cesswairimu  Can you (or any of the mentors!) check my code to make sure I'm doing this correctly? Thank you!!

cesswairimu commented:
Sure, not sure why travis is failing....taking a look

nstjean commented:
@jywarren Can you take a look at my code and let me know if this is what you intended? I wasn't sure if you wanted to reference the form at all in the Javascript.

nstjean commented:
AddComment is done!

I had to rebase in branch before pushing, I don't know if I did it right. Now it's showing me a bunch of codeclimate errors.

jywarren commented:
Wow, it looks like this may need some rebasing and cleaning up! @nstjean this can be a bit rough, but please save a local copy of your branch with `git branch refactor-js-backup` so nothing goes wrong. 

Would you like to try rebasing again over the latest master branch? I can also help a bit with this!

There's some guidance here, however!

https://publiclab.org/wiki/contributing-to-public-lab-software#Rewinding+the+master+branch

jywarren commented:
Also, i think this needs parentheses to disambiguate the `||` logic, just a note!

![image](https://user-images.githubusercontent.com/24359/69174707-51cbaf80-0ad0-11ea-8364-1e415796026c.png)

The rest of the CodeClimate changes can be ignored for now, as they're mostly stylistic!

jywarren commented:
Thanks for sticking with this one! It's a really awesome one! Just ping me if you get super stuck!

nstjean commented:
Yes I'll try rebasing using rewinding the master branch!

nstjean commented:
@jywarren I reset the HEAD, switched to my branch, rebased. But now when trying to run tests on my local everything is failing, even in Master. Do I have to do something else?

![Screenshot Test Errors](https://user-images.githubusercontent.com/49460529/69360843-48218380-0c59-11ea-8a24-0c6abfdf74bb.png)

nstjean commented:
When I try to run my test for my javascript function I get this error, which points to a Chrome version error. But I haven't changed anything other than pulling the new version of Master.
```
ERROR["test_adding_a_tag_via_javascript", #<Minitest::Reporters::Suite:0x00005587bc60d350 @name="CommentTest">, 1.5917023420042824]
 test_adding_a_tag_via_javascript#CommentTest (1.59s)
Selenium::WebDriver::Error::SessionNotCreatedError:         Selenium::WebDriver::Error::SessionNotCreatedError: session not created: This version of ChromeDriver only supports Chrome version 78
            test/system/comment_test.rb:9:in `block in <class:CommentTest>'
```

jywarren commented on app/assets/javascripts/ajax.js:
```diff
@@ -0,0 +1,9 @@
+function sendFormSubmissionAjax(dataObj, selector) {
```
This is looking good! Perhaps this file needs a more descriptive name, however; but is it a generic utility? If so, maybe can add a sentence or two on how to use it?

Thanks!

nstjean replied:
Yes, I should do that!

jywarren commented:
This looks good!

Not sure where this error is coming from though:

```
ERROR["test_find_by_tagname_and_user_id", #<Minitest::Reporters::Suite:0x00005653ad056bb0 @name="NodeTest">, 2.091748925999923]
 test_find_by_tagname_and_user_id#NodeTest (2.09s)
NoMethodError:         NoMethodError: undefined method `title' for nil:NilClass
            test/unit/node_test.rb:529:in `block in <class:NodeTest>'
```

That's for this test:

https://github.com/publiclab/plots2/blob/8c2713fa10bdd023f193e3957f2f047560f20e91/test/unit/node_test.rb#L529

... i guess it seems unrelated? I'll try restarting the unit tests to see what happens..

Thanks for the epic rebase!!!!

jywarren commented on test/system/tag_test.rb:
```diff
@@ -0,0 +1,27 @@
+require 'test_helper'
```
Wow, really awesome tests!!!!

jywarren commented on test/system/comment_test.rb:
```diff
@@ -0,0 +1,26 @@
+    page.evaluate_script("addComment('fantastic four')")
+
+    # check that the tag showed up on the page
+    assert_selector('#comments-list .comment-body p', text: 'fantastic four')
```
🎉

jywarren commented on app/assets/javascripts/submit_form_ajax.js:
```diff
@@ -0,0 +1,12 @@
+// Allows data to be submitted from anywhere on the page using Javascript without using the form itself
+function sendFormSubmissionAjax(dataObj, selector) {
+  $.ajax({
+    url: $(selector).attr('action'), // grab the URL from the form
```
Could we offer a parameter like `url` which defaults back to this value? That way it could be done without a form!

jywarren commented:
Could we... Refactor the form so it submits through ajax only? I am not
that worries about the ambiguity but I would be fine consolidating
everything into a single remote call a dropping the direct use of the html
form if it's bothering you! Thanks!

On Tue, Nov 26, 2019, 3:31 PM Natalie St Jean <notifications@github.com>
wrote:

> @jywarren <https://github.com/jywarren> I've been thinking about this.
> Will it get confusing if sometimes the function is passed a URL and
> sometimes it is passed the ID of the form? It only needs one or the other.
> Maybe it should be two different functions? Say function2 takes in a URL as
> a parameter and submits the form. Then function2 takes in a selector, gets
> the URL from the form, and then calls function1?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/6422?email_source=notifications&email_token=AAAF6J5NGAI7YGZC2CI4TJLQVWBPTA5CNFSM4I6U7RQKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFHK65A#issuecomment-558804852>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J4YCOYPN6BJHAP6P7DQVWBPTANCNFSM4I6U7RQA>
> .
>

nstjean commented:
Okay! :)

nstjean commented:
Ok I allowed either a selector or a url to get passed in and it decides what to do from there. :) I also added a test case for a URL being used in both addComment and addTag. Tests work on local.

jywarren commented on app/assets/javascripts/comment.js:
```diff
@@ -51,3 +51,10 @@ function insertTitleSuggestionTemplate() {
 }
+
+// JS API for submitting comments
+function addComment(comment, submitTo) {
```
Great, and this works appropriately for nested comments (responding to other comments) as well as top-level comments? Just checking!

jywarren commented on test/system/tag_test.rb:
```diff
@@ -0,0 +1,46 @@
+
+  end
+
+  test 'adding a tag via javascript with url only' do
```
Love this. Just one more question - does adding tags to your profile also still work? I think that might've become broken at some point because we (unfortunately) share code between node tags and user tags, but just curious if this touches that at all? Thanks!

nstjean commented:
Let me check on those possibilities!

jywarren commented:
Thanks. What's really nice about your code is that we will be moving away from more brittle implementations like the current one, and backing it up with tests, so the hope is to have fewer "gotchas" like this in the future. Thanks a lot!!!

jywarren commented:
For what it's worth, i just tried adding a profile tag and got some kind of error i think... it submitted via XHR (ajax) to: https://publiclab.org/profile/warren?remote=true&name=providence also which I think isn't right!

jywarren commented:
So, what we can do is merge this, pending the comment-responses clarification above, and then in a new PR (if you're willing!) try to fix the profile tagging -- preferably by extending the methods you've written, to have a more rational system for handling profile tagging, too! 

The nice thing is that we are trying to add profile tags at different places around the site, so having a general purpose JS method to do this will be really great.

nstjean commented:
Ok! I added the ability to add a parent comment ID number to the function call so it will nest it appropriately... and added a test to check.

I can check on the profile tag implementation in another PR!

jywarren commented:
Awesome ok this is ready to merge then?

On Wed, Nov 27, 2019, 7:36 PM plotsbot <notifications@github.com> wrote:

> 1 Warning
> ⚠️ There was an error with Danger bot’s Junit parsing: No JUnit file was
> found at output.xml
> 3 Messages
> 📖 @nstjean <https://github.com/nstjean> Thank you for your pull request!
> I’m here to help with some tips and recommendations. Please take a look at
> the list provided and help us review and accept your contribution! And *don’t
> be discouraged if you see errors* – we’re here to help.
> 📖 It looks like you haven’t marked all the checkboxes. Help us review
> and accept your suggested changes by going through the steps one by one. If
> it is still a ‘Work in progress’, please include ‘[WIP]’ in the title.
> 📖 # Screenshots 📸 (click to expand)6422-test_questions.png
>
>
> <https://camo.githubusercontent.com/b4cfc256f99a3896732b48ec2f3b79f103ddddec/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7175657374696f6e732e706e67>
> 6422-test_embeddable_grids.png
>
>
> <https://camo.githubusercontent.com/1a29d69d10db4f805bd196b925e1c94a4d8b25d5/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f656d6265646461626c655f67726964732e706e67>
> 6422-test_signup.png
>
>
> <https://camo.githubusercontent.com/f9057bd550885ff0bc9d3d47030013cee66180e9/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7369676e75702e706e67>
> 6422-test_viewing_the_settings_page.png
>
>
> <https://camo.githubusercontent.com/867d894242847090282d3724d3a9111b18b67b93/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f76696577696e675f7468655f73657474696e67735f706167652e706e67>
> 6422-test_tag_by_author_page.png
>
>
> <https://camo.githubusercontent.com/b86cb57ea396d4202775de78c4ccda6212d337fd/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7461675f62795f617574686f725f706167652e706e67>
> 6422-test_wiki_page_with_inline_grids.png
>
>
> <https://camo.githubusercontent.com/19c97950efe7aae0d4219a4fea52c7e7f22718a1/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f77696b695f706167655f776974685f696e6c696e655f67726964732e706e67>
> 6422-test_stats.png
>
>
> <https://camo.githubusercontent.com/cbc09cc89982b513b6200dd99dfb75996fa6b4ca/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f73746174732e706e67>
> 6422-test_viewing_the_dashboard.png
>
>
> <https://camo.githubusercontent.com/1fc08cb82e6c0f1b84aab6a361fae2cd3a90234b/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f76696577696e675f7468655f64617368626f6172642e706e67>
> 6422-test_searching_an_item_from_the_homepage.png
>
>
> <https://camo.githubusercontent.com/4a8e9e87dbcb0c9b3421ed1ee527a72cd4bed75e/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f736561726368696e675f616e5f6974656d5f66726f6d5f7468655f686f6d65706167652e706e67>
> 6422-test_questions_shadow.png
>
>
> <https://camo.githubusercontent.com/e9192edc452d4af588237aa80df3ed243001b689/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7175657374696f6e735f736861646f772e706e67>
> 6422-test_login_modal.png
>
>
> <https://camo.githubusercontent.com/db46860b8f00174f7d696141b435ed3e933f9cf7/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f6c6f67696e5f6d6f64616c2e706e67>
> 6422-test_profile_page.png
>
>
> <https://camo.githubusercontent.com/fb1a367c75bbd090203b2b10bc7d90c9d595c387/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f70726f66696c655f706167652e706e67>
> 6422-test_comments.png
>
>
> <https://camo.githubusercontent.com/b18403f3a109714a7236cdfe46535e7acadba246/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f636f6d6d656e74732e706e67>
> 6422-test_tags.png
>
>
> <https://camo.githubusercontent.com/050bfb422806dc4f96f209680bfcc49730d6bc79/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f746167732e706e67>
> 6422-test_signup_modal.png
>
>
> <https://camo.githubusercontent.com/16043354c74b207254c5d9d65f73760acd5fa76d/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7369676e75705f6d6f64616c2e706e67>
> 6422-test_wiki.png
>
>
> <https://camo.githubusercontent.com/e5af982c3fbd2195f077097eaa3c3bb289d71b82/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f77696b692e706e67>
> 6422-test_methods.png
>
>
> <https://camo.githubusercontent.com/839305666f8e86f294e980c1f7a3838b8e699a4b/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f6d6574686f64732e706e67>
> 6422-test_tag_page.png
>
>
> <https://camo.githubusercontent.com/82c1fb1b3cf8fc71311e5fbc3cc3aaa464e4ed98/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7461675f706167652e706e67>
> 6422-test_blog_page_with_location_modal.png
>
>
> <https://camo.githubusercontent.com/ad07f106e820abdf72fe51140487be0aaf049caa/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f626c6f675f706167655f776974685f6c6f636174696f6e5f6d6f64616c2e706e67>
> 6422-test_tag_wildcard.png
>
>
> <https://camo.githubusercontent.com/ecaa60ee9fbf4c688889d7ed5a8b97444a6debef/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7461675f77696c64636172642e706e67>
> 6422-test_embeddable_thumbnail_grids.png
>
>
> <https://camo.githubusercontent.com/85157b87eba68027ce60b5eb5a496b425c1bf39b/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f656d6265646461626c655f7468756d626e61696c5f67726964732e706e67>
> 6422-test_front_page_with_navbar_search_autocomplete.png
>
>
> <https://camo.githubusercontent.com/4ebb596a3b2b42343129f185a22ce1bd8a4709a3/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f66726f6e745f706167655f776974685f6e61766261725f7365617263685f6175746f636f6d706c6574652e706e67>
> 6422-test_login.png
>
>
> <https://camo.githubusercontent.com/c6ea2d08c247f882b2581af04840ebc3b7b737c4/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f6c6f67696e2e706e67>
> 6422-test_viewing_the_dropdown_menu.png
>
>
> <https://camo.githubusercontent.com/954cf28a968f5b4683894c21178dbb99b1c17de1/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f76696577696e675f7468655f64726f70646f776e5f6d656e752e706e67>
> 6422-test_viewing_question_post.png
>
>
> <https://camo.githubusercontent.com/8eb7792d3822d6a0af1dea736525a6dfa75a91c3/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f76696577696e675f7175657374696f6e5f706f73742e706e67>
> 6422-test_mobile_displays.png
>
>
> <https://camo.githubusercontent.com/836b57999b160505fff62eecdfb144d7049690fc/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f6d6f62696c655f646973706c6179732e706e67>
> 6422-test_simple-data-grapher_powertag.png
>
>
> <https://camo.githubusercontent.com/80cd068cc4ad5c4e73732f5552c2bec2c056420f/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f73696d706c652d646174612d677261706865725f706f7765727461672e706e67>
> 6422-test_front.png
>
>
> <https://camo.githubusercontent.com/ab20be2aa7b6f54cd5656f6b79b91f6dea438299/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f66726f6e742e706e67>
> 6422-test_question_page.png
>
>
> <https://camo.githubusercontent.com/bb4b46492afb41a7db6fbfdb45cebc7d2bd90c64/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7175657374696f6e5f706167652e706e67>
> 6422-test_tag_contributors_page.png
>
>
> <https://camo.githubusercontent.com/468dd129ead30e641baf4842278c3def22ba6ee6/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f7461675f636f6e7472696275746f72735f706167652e706e67>
> 6422-test_blog.png
>
>
> <https://camo.githubusercontent.com/93a532656b0d637d434c4314253d89488e1c42c9/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f626c6f672e706e67>
> 6422-failures_test_adding_a_tag_via_javascript_with_url_only.png
>
>
> <https://camo.githubusercontent.com/66e7a23fbf9749c94b19417492e8070ec3b87441/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d6661696c757265735f746573745f616464696e675f615f7461675f7669615f6a6176617363726970745f776974685f75726c5f6f6e6c792e706e67>
> 6422-test_people.png
>
>
> <https://camo.githubusercontent.com/fe7736eb60e0c9979f919ac311605f170dfb0201/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f70656f706c652e706e67>
> 6422-test_wiki_revisions.png
>
>
> <https://camo.githubusercontent.com/7428171f0f8817fb4210ff994843030a7e81c79c/68747470733a2f2f73746f726167652e676f6f676c65617069732e636f6d2f706c6f7473322d73637265656e73686f74732f363432322d746573745f77696b695f7265766973696f6e732e706e67>
>
> Learn about automated screenshots
> <https://github.com/publiclab/plots2/issues/5316>
>
> Generated by 🚫 Danger <https://danger.systems/>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/6422?email_source=notifications&email_token=AAAF6J2O7Y5MR6Z6OBAGP7DQV4G75A5CNFSM4I6U7RQKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFLDCWY#issuecomment-559296859>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J7OTS4AKG4KWXVLXYLQV4G75ANCNFSM4I6U7RQA>
> .
>

nstjean commented:
It's throwing an error on my test test_adding_a_tag_via_javascript_with_url_only - but it works when I run it on local. I'll try to figure out what's going on with it.

nstjean commented:
It appears I have fixed it! Should be all set!

jywarren commented:
Super! Now go take a break! 😁🎉🙌🏽

You'll be able to test it out at stable.publiclab.org soon once it builds!

nstjean commented:
Thanks! :tada: Enjoy your Thanksgiving break!
